Source/NSData.m

```diff
@@ -1082,6 +1082,119 @@ - (NSData*) subdataWithRange: (NSRange)aRange
   return [NSData dataWithBytesNoCopy: buffer length: aRange.length];
 }
 
+/**
+ * Finds and returns the range of the first occurrence of the given data, within the given range, subject to given options.
+ */
+- (NSRange) rangeOfData: (NSData *)dataToFind
+                options: (NSDataSearchOptions)mask
+                  range: (NSRange)searchRange
+{
+  NSUInteger  length = [self length];
+  NSUInteger  countOther = [dataToFind length];
+  const void* bytesSelf = [self bytes];
+  const void* bytesOther = [dataToFind bytes];
+  NSRange     result;
+
+  GS_RANGE_CHECK(searchRange, length);
+  if (dataToFind == nil)
+    [NSException raise: NSInvalidArgumentException format: @"range of nil"];
+
+  /* Zero length data is always found at the start of the given range.
+   */
+  if (0 == countOther)
+    {
+      if ((mask & NSDataSearchBackwards) == NSDataSearchBackwards)
+        {
+          searchRange.location += searchRange.length;
+        }
+      searchRange.length = 0;
+      return searchRange;
+    }
+
+  if (searchRange.length < countOther)
+    {
+      /* Range to search is smaller than data to look for.
+       */
+      result = NSMakeRange(NSNotFound, 0);
+    }
+  else
+    {
+      if ((mask & NSDataSearchAnchored) == NSDataSearchAnchored
+        || searchRange.length == countOther)
+        {
+          /* Range to search is same size as data to look for.
+           */
+          if ((mask & NSDataSearchBackwards) == NSDataSearchBackwards)
+            {
+              searchRange.location = NSMaxRange(searchRange) - countOther;
+              searchRange.length = countOther;
+            }
+          else
+            {
+              searchRange.length = countOther;
+            }
+          if (memcmp(&bytesSelf[0], &bytesOther[0], countOther) == 0)
```

Comment:
Why not `memcmp(bytesSelf, &bytesOther, countOther)`?

I think that this is also wrong as the anchor search could start from the end, so we should not just compare the start of self.

Reply:
The indices were just specified because that's how it was done in [NSString rangeOfString:options:range], but I agree it would be clearer to change it.

I'm not quite sure it's wrong yet though, unless rangeOfString is wrong too.  Certainly I could have made a mistake and I'll add some tests to try to work out how it should behave with both NSDataSearchAnchored and NSDataSearchBackwards, but after a quick review it still appears to me to be working the same as rangeOfString.

Thank you!

Reply:
I never reviewed the code in NSString. A quick look shows that it probably used the indices to keep it similar to the other places where the compare starts from a specific index. I must admit consistency is good enough a reason to do it this way.

Reply:
> The indices were just specified because that's how it was done in ...
I think there is a problem in that the code here, not in that it uses indices, but that it uses the wrong indices.
In NSString.m the indices used are zero because the relevant range of characters has been extracted from the string and placed in new unichar buffers, and we are doing a comparison of those unichar sequences.
In NSData.m bytesSelf and bytesOther are pointers to the start of the raw data, so the correct offsets must be used.  For the data being checked with an anchored search, that offset is searchRange.location which may not be zero.